packages/nuxt/src/pages/plugins/route-rules.ts

```diff
@@ -0,0 +1,56 @@
+import { runInNewContext } from 'node:vm'
+import { pathToFileURL } from 'node:url'
+import { createUnplugin } from 'unplugin'
+import type { Node } from 'estree-walker'
+import type { CallExpression } from 'estree'
+import { walk } from 'estree-walker'
+import { normalize } from 'pathe'
+import { parseURL } from 'ufo'
+
+const ROUTE_RULE_RE = /\bdefineRouteRules\(/
+
+interface RouteRuleExtractorPluginOptions {
+  routeRules: Record<string, any>
+  pageMap: Record<string, string>
+  updateConfig: (() => void | Promise<void>)
+}
+
+export const routeRuleExtractorPlugin = createUnplugin((options: RouteRuleExtractorPluginOptions) => {
+  return {
+    name: 'test',
+    transformInclude (id) {
+      const { pathname } = parseURL(decodeURIComponent(pathToFileURL(normalize(id)).href))
+      return pathname in options.pageMap
+    },
+    async transform (code, id) {
+      if (!ROUTE_RULE_RE.test(code)) { return null }
+      const { pathname } = parseURL(decodeURIComponent(pathToFileURL(normalize(id)).href))
+
+      let rule = {}
+
+      walk(this.parse(code, {
+        sourceType: 'module',
+        ecmaVersion: 'latest'
+      }) as Node, {
```

Comment:
Is it possible to share one plugin for pageMeta and avoid parsing twice?

Reply:
I've recently updated the page meta plugin so it only processes files with the query string, which shouldn't be processed with this plugin.

But the content will be nearly identical.

I wonder if there's a way to cache the AST?